### DISCUSSION TITLE
Why does ndarray_wrap tend to copy?

### DISCUSSION THREAD

**Discussion by WKarel** [wjakob/nanobind#188] [General]
Given an ndarray_handle and numpy as framework, ndarray_wrap creates a new nb_ndarray instance (that shares the data of the ndarray_handle), and calls numpy.array(nb_ndarray), which makes NumPy request a buffer from the nb_ndarray, again sharing the data. All reference counting seems fine, and no copying of data is involved if e.g. none is passed as return value policy.

With the defaults passed by nb::cast (copy/move), the data IS copied, however. What is the reason for that, given that ndarray_handle et al. have all means to share data? Why does ndarray_wrap consider a return value policy at all? This seems to contradict the documentation saying "A return value policy is unnecessary when the type itself clarifies ownership...".

Thanks!

**Comment by wjakob** [comment]
For historical context: Initially, nanobind would make a DLPack capsule and then convert that into an nd-array of choice (NumPy, PyTorch, etc.). However, the NumPy <-> DLpack interface was found to be insufficient since non-brand-new numpy versions didn't support it, so I added an extra indirect path via an `nb_ndarray` exposing the buffer protocol. Later, people noticed that the NumPy <-> DLPack interface is too limited even on brand-new NumPy versions, so I entirely switched it over to the buffer protocol and disabled the DLPack code path for NumPy.

Can you explain the problem you are experiencing with the return value policies? It may be helpful to see an actual example.

**Reply by WKarel** [reply]
Thanks for these insights!

See the example below:
1. `numpy_ndarray` directly returns a `nb::ndarray`. Nanobind casts this to a `numpy.ndarray` that shares the data.
2. `numpy_ndarray_cast_default` explicitly casts the `nb::ndarray` with default arguments and hence returns a `numpy.ndarray`. But the data is copied in this case.
3. `numpy_ndarray_cast_none` explicitly casts the nb::ndarray with `none` as return value policy. This avoids copying the data. `none` seems to be misused here, but given the implementation of infer_policy, it is the only policy that avoids a copy, apart from `take_ownership`, .

To my understanding, nanobind ignores return value policies for types that support shared ownership. Please correct me if I'm wrong. Since nb::ndarray/nb::ndarray_handle/buffer/numpy.ndarray all have this feature, I would expect all 3 functions to return the original data. Even more, since the elaborate implementation seems to work fine.

Looking at the implementation of `ndarray_wrap`, [this](https://github.com/wjakob/nanobind/blob/ec9350b805d2fe568f65746fd69225eedc5e37ae/src/nb_ndarray.cpp#L546) line called my attention, which seems to explicitly force copies in many situations:
`bool copy = policy == rv_policy::copy || policy == rv_policy::move;`

As said, I find it surprising that `ndarray_wrap` considers a return value policy at all.

1. calls `ndarray_wrap` with `policy = rv_policy::automatic`,
2. with `policy = rv_policy::move`,
3. with `policy = rv_policy::none`

C++ code:
```
#include <nanobind/nanobind.h>
#include <nanobind/ndarray.h>
#include <sstream>

namespace nb = nanobind;

nb::ndarray<double, nb::numpy> nparray()
{
  size_t shape = 3;
  double* data = new double[shape];
  for (size_t idx = 0; idx < shape; ++idx)
    data[idx] = double(idx);
  std::ostringstream oss;
  oss << std::hex << std::showbase << (std::uintptr_t)data;
  nb::print(oss.str().c_str());
  nb::capsule owner{ data, [](void* memory) noexcept { delete[] (double*)memory; } };
  return {data, 1, & shape, owner.ptr()};
}

NB_MODULE(nanobind_zerocopy, m)
{
  m.def("numpy_ndarray", []()
    {
      return nparray();
    });

  m.def("numpy_ndarray_cast_default", []()
    {
      return nb::cast(nparray());
    });

  m.def("numpy_ndarray_cast_none", []()
    {
      return nb::cast(nparray(), nb::rv_policy::none);
    });
}
```

Python code:
```
import nanobind_zerocopy

print('numpy_ndarray')
arr = nanobind_zerocopy.numpy_ndarray()
print(hex(arr.ctypes.data), arr, arr.flags, sep='\n')

print('\nnumpy_ndarray_cast_default')
arr = nanobind_zerocopy.numpy_ndarray_cast_default()
print(hex(arr.ctypes.data), arr, arr.flags, sep='\n')

print('\nnumpy_ndarray_cast_none')
arr = nanobind_zerocopy.numpy_ndarray_cast_none()
print(hex(arr.ctypes.data), arr, arr.flags, sep='\n')
```

Running NumPy 1.24.2 on Windows, this prints e.g.:
```
numpy_ndarray
0x1a6ed2c1780
0x1a6ed2c1780
[0. 1. 2.]
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False


numpy_ndarray_cast_default
0x1a6ff297050
0x1a6ff2f0a50
[0. 1. 2.]
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False


numpy_ndarray_cast_none
0x1a6ff296510
0x1a6ff296510
[0. 1. 2.]
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
```

**Reply by wjakob** [reply]
Hi @WKarel,

sorry about the delay, I missed the notification about your message.

Return value policies were originally created to provide additional context when returning/casting types bound using `class_<..>`. They can be used in more complex constructions (e.g. STL maps of classes), but are in that case still passed downwards for use in the central class type caster.

The use of type casters in something like `nb::ndarray` is sort of a gray zone, and the statements about shared ownership IMO don't apply there directly.

For example, it is very useful (and sometimes necessary) to create a copy of a tensor. Doing so in a generic way can be difficult given the plethora of different frameworks that can create them. So it's a useful feature that the ndarray type caster has the ability to copy an array if needed, and return value policies seem like the obvious way to expose such functionality.

Now, the details are a bit more subtle -- should ``nb::cast(..)`` always force a copy of an ndarray? I have actually never done this before (since ndarrays can be directly returned), and I agree that the behavior is surprising.

Best,
Wenzel

**Reply by WKarel** [reply]
Dear @wjakob ,

I think that my use case is quite basic: bind a function that returns a `nb::list` filled with newly created instances of `nb::ndarray`s (prvalues as e.g. returned from my function `nparray` above; assume that they are big, such that their data are expensive to copy):

```
m.def("numpy_ndarrays_cast_default", []()
    {
      nb::list li;
      li.append(nparray());
      return li;
    });
```
For some reason, the current implementation of `nb::list::append` [does not](https://github.com/wjakob/nanobind/blob/3883e28bec0fd5ea539eb88ca389ecac52d9bfa8/include/nanobind/nb_cast.h#LL426C21-L426C21) use `nb::detail::forward<T>` to forward its argument when calling `nb::cast`. Hence, `nb::detail::infer_policy<T>(automatic_reference)` within `nb::cast` deduces `copy` as return value policy for the call to `type_caster<ndarray<...>>::from_cpp` -> `ndarray_wrap`, which then creates a `nb::ndarray` with a copy of the data, even though the passed `nb::ndarray` was only created for that call. 

If `nb::cast` is called with `move` as return value policy (either explicitly, or because `list::append` has been adapted to forward its argument), then `ndarray_wrap` gets called with this return value policy. But even then, [said line](https://github.com/wjakob/nanobind/blob/3883e28bec0fd5ea539eb88ca389ecac52d9bfa8/src/nb_ndarray.cpp#L625) in `ndarray_wrap` forces a copy:
```
m.def("numpy_ndarrays_cast_none", []()
    {
      nb::list li;
      li.append(nb::cast(nparray(), nb::rv_policy::move));
      return li;
    });
```

Currently, the only way to avoid copying the data is to misuse a return value policy like `none`, which seems like a bug in application code, since no respective PyObject exists yet!

I suggest the following:
- change the signature of `ndarray_wrap` to expect `bool copy` instead of `rv_policy`.
- remove [this line](https://github.com/wjakob/nanobind/blob/3883e28bec0fd5ea539eb88ca389ecac52d9bfa8/src/nb_ndarray.cpp#L625).
- define the method `ndarray::copy` to explicitly copy an ndarray, with an `ndarray_framework` as argument (that defaults to the one of `ndarray`), calling `ndarray_wrap` with `copy=true`, followed by `ndarray_import` (this might not be the most efficient approach to achieve a copy).
- make `type_caster<ndarray<...>>` ignore its return value policy and call `ndarray_wrap` with `copy=false`.
- make `list::append` forward its argument using `detail::forward_t<T>`.

By the way: `ndarray_wrap` ignores the return value policy if `ndarray_handle::self` is set. In this case, it does not only not copy by default, but there is even no way to make it do so - which is almost the opposite of the behaviour if `ndarray_handle::self` is unset.

Cheers,
Willi.

**Reply by wjakob** [reply]
> For some reason, the current implementation of nb::list::append [does not](https://github.com/wjakob/nanobind/blob/3883e28bec0fd5ea539eb88ca389ecac52d9bfa8/include/nanobind/nb_cast.h#LL426C21-L426C21) use nb::detail::forward<T> to forward its argument when calling nb::cast.

That's an oversight. Fixed in 2219d0b0fec5e6cc4fce96bc3dbad6bfa148a57d.

> By the way: ndarray_wrap ignores the return value policy if ndarray_handle::self is set. In this case, it does not only not copy by default, but there is even no way to make it do so - which is almost the opposite of the behaviour if ndarray_handle::self is unset.

Fixed in https://github.com/wjakob/nanobind/commit/ffd22b069ba95a546baeca0bdb6711fb9059cad8.

I'm not a fan of the `.copy` method you proposed, which seems inefficient and makes the API more cumbersome. And completely disabling copying the caster also seems bug-prone. I will think a bit more about it.

**Reply by wjakob** [reply]
Ok, here is my take on this issue:

1. First, let's get rid of the scattered ``nb::detail::infer_policy`` calls, which are problematic (see commit message): https://github.com/wjakob/nanobind/commit/1ff9df03fb56a16f56854b4cecd1f388f73d3b53. That way, the `ndarray` type caster will usually get `rv_policy::automatic` or `rv_policy::automatic_reference` depending on whether or not it is used in a function return value or `nb::cast`.

2. Implement the (almost) full set of return value policies in `nb::ndarray`: https://github.com/wjakob/nanobind/commit/a79575165134c72c0a26e46772290d0404eae7a3

3. Minor unrelated quality of life improvement: https://github.com/wjakob/nanobind/commit/de111766b21fe893a41cd4614a346b0da251f7f2

**Reply by wjakob** [reply]
Can you let me know if this looks reasonable to you?

**Reply by WKarel** [reply]
My use case works now, and so I'm happy, thanks. Still, a dedicated copy-method would seem cleaner to me from an application's point of view, with `type_caster` treating `ndarray` like an intrusive pointer for `detail::ndarray_handle`. But maybe I miss an obstacle in there.

Besides, I have posted 3 comments for the commits you mentioned above.

**Reply by wjakob** [reply]
>  treating ``ndarray`` like an intrusive pointer for ``detail::ndarray_handle``. 

Isn't that what it does already? (including, e.g., the necessary reference counting)

Copying the array is quite entangled with actual type casting step. For many array types (e.g. a CUDA or hypothetical ROCm array) we don't have access (at least directly in nanobind) to the API that would let us copy the array, and we need to ask the framework to do this when converting from an abstract device pointer to a Python object. Hence the reason to package this up as a return value policy

**Reply by wjakob** [reply]
Thanks for the three comments, I incorporated them.

**Reply by WKarel** [reply]
>> treating ndarray like an intrusive pointer for detail::ndarray_handle.

> Isn't that what it does already? (including, e.g., the necessary reference counting)

I meant the behaviour that e.g. `type_caster<std::shared_ptr>::from_cpp` has: it simply ignores the passed return value policy, and will never copy the pointee. But never mind, I see that the implementation is simpler this way.
